### ISSUE TITLE
Evaluation of k-means for clustering

### ISSUE THREAD

**Issue by Ladas** [RedHatInsights/aicoe-insights-clustering#18]
Plotting SSE and silhouette coefficient (multiplying by 3M, so we can see it in 1 chart)

![image](https://user-images.githubusercontent.com/1737058/46600157-945b4400-cae9-11e8-988b-e0e8b85b319f.png)

----------

Trying to see what variance is in the features, we can that around 245 features hold 99% of the variance

![image](https://user-images.githubusercontent.com/1737058/46600231-d5535880-cae9-11e8-8baa-1e03b88c7e8d.png)

---------------------------

Lets try to see SSE and silhouette coefficient when having only 245 components, transformed by PCA 

![image](https://user-images.githubusercontent.com/1737058/46600312-10558c00-caea-11e8-9d96-2ef7b029cdc1.png)

-----------------------------

Result: we can see that max silhouette coefficient is still around 0.43, on just 4 clusters, which has bad SSE. Then it stays around 0.3.

What sources say for silhouette coefficient in that area ```0.26-0.50	The structure is weak and could be artificial```

Which points to that we should try a different clustering method, since k-means seems to have bad score.

**Comment by Ladas** [comment]
cc @durandom @tumido @MichaelClifford does it make sense to you? The silhouette coefficient shows that k-means is probably not a good method in this case.

**Comment by durandom** [comment]
Maybe this is also because of the input data we are using.
I've written up the next steps proposed [here](https://docs.google.com/document/d/1INXfpGjjMRsrE_fnY3t0U3btTYhZWuPPKCsAaDeToDc/edit#heading=h.o8d6mhpl9qm)

**Comment by durandom** [comment]
Is this a measurement you could integrate into the [metric_tracking package](https://github.com/RedHatInsights/aicoe-insights-clustering/pull/14/files#diff-d0301332bd6fef353ec35837646aa49e)
Whats the input of the silhouette coefficient? Just the clusters? Could you add the notebook or code to produce this?

**Comment by Ladas** [comment]
@durandom yes, I plan to add it there once it is merged. Also, I'll need to figure out how to run it as separate jobs, now it takes hours to compute for 1 day on 4 cores.

Using the rule data from 2018-09-05 as input

**Comment by durandom** [comment]
> Also, I'll need to figure out how to run it as separate jobs, now it takes hours to compute for 1 day on 4 cores.

Thats what the upshift environment is meant for. Create a build config and let it run there.

**Comment by tumido** [comment]
@Ladas thank you for the metrics! It makes total sense and just proves that we all know, that we know nothing. :smile:

There are probably multiple factors in effect:
- data preprocessing (ironing out the structure too much)
- applied clustering metod which doesn't fit the use case
- input data composition (extracting wrong markers, missing important attributes, etc.)

I think this is the kind of insight into the clustering Marcel was looking for and these graphs would make it easier to compare different solutions. I'd love to see them as a part of the metrics tracking thingy Marcel's team is working on. :+1:

**Comment by Ladas** [comment]
FYI. seems like DBscan also performs bad, we'll need to munge the data before clustering

![image](https://user-images.githubusercontent.com/1737058/46654703-d8158280-cba9-11e8-906e-3ee6910e62f5.png)

![image](https://user-images.githubusercontent.com/1737058/46654718-e6639e80-cba9-11e8-825b-f12db7257b57.png)

**Comment by durandom** [comment]
cc @TreeinRandomForest
